ServerInterface::PayloadAsyncRequest does not work with ByteBuffer

### Should this be an issue in the gRPC issue tracker?
 Yes
 
### What version of gRPC and what language are you using?
 gRPC 1.10; C++
 
### What operating system (Linux, Windows, …) and version?
 Windows 10 (1709)
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
 Visual Studio 2015
 
### What did you do?
Updated [`grpc::SerializationTraits<bond::bonded<T>>::Deserialize`](https://github.com/Microsoft/bond/blob/master/cpp/inc/bond/ext/grpc/bond_utils.h#L60) to use `grpc::ByteBuffer` instead of `grpc_byte_buffer`.
 
### What did you expect to see?
 No compilation errors.
 
### What did you see instead?
 Compilation error at [`grpc::ServerInterface::PayloadAsyncRequest<Message>::FinalizeResult`](https://github.com/grpc/grpc/blob/master/include/grpcpp/impl/codegen/server_interface.h#L194), since it expects `grpc_byte_buffer*` instead of `grpc::ByteBuffer*`.
 
### Anything else we should know about your project / environment?

Here is the actual error:
`E:\depot\bond\thirdparty\grpc\include\grpcpp/impl/codegen/server_interface.h(194): error C2664: 'grpc::Status grpc::SerializationTraits<R,void>::Deserialize(grpc::ByteBuffer *,bond::bonded<bond::Void,bond::ProtocolReader> *)': cannot convert argument 1 from 'grpc_byte_buffer *' to 'grpc::ByteBuffer *'`.

--- a/include/grpcpp/impl/codegen/byte_buffer.h
+++ b/include/grpcpp/impl/codegen/byte_buffer.h
@@ -31,6 +31,8 @@
 
 namespace grpc {
 
+class ServerInterface;
+
 namespace internal {
 class CallOpSendMessage;
 template <class R>
@@ -98,6 +100,7 @@ class ByteBuffer final {
 
  private:
   friend class SerializationTraits<ByteBuffer, void>;
+  friend class ServerInterface;
   friend class internal::CallOpSendMessage;
   template <class R>
   friend class internal::CallOpRecvMessage;

--- a/include/grpcpp/impl/codegen/server_interface.h
+++ b/include/grpcpp/impl/codegen/server_interface.h
@@ -186,7 +186,11 @@ class ServerInterface : public internal::CallHook {
           notification_cq_(notification_cq),
           tag_(tag),
           request_(request) {
-      IssueRequest(registered_method, payload_.bbuf_ptr(), notification_cq);
+      IssueRequest(registered_method, payload_.c_buffer_ptr(), notification_cq);
+    }
+
+    ~PayloadAsyncRequest() {
+        payload_.Release(); // We do not own the payload_
     }
 
     bool FinalizeResult(void** tag, bool* status) override {

--- a/include/grpcpp/impl/codegen/server_interface.h
+++ b/include/grpcpp/impl/codegen/server_interface.h
@@ -190,14 +190,14 @@ class ServerInterface : public internal::CallHook {
     }
 
     ~PayloadAsyncRequest() {
-        payload_.Release(); // We do not own the payload_
+      payload_.Release();  // We do not own the payload_
     }
 
     bool FinalizeResult(void** tag, bool* status) override {
       if (*status) {
-        if (!payload_.Valid() ||
-            !SerializationTraits<Message>::Deserialize(payload_.bbuf_ptr(), request_)
-                 .ok()) {
+        if (!payload_.Valid() || !SerializationTraits<Message>::Deserialize(
+                                      payload_.bbuf_ptr(), request_)
+                                      .ok()) {
           // If deserialization fails, we cancel the call and instantiate
           // a new instance of ourselves to request another call.  We then
           // return false, which prevents the call from being returned to

--- a/include/grpcpp/impl/codegen/server_interface.h
+++ b/include/grpcpp/impl/codegen/server_interface.h
@@ -20,6 +20,7 @@
 #define GRPCPP_IMPL_CODEGEN_SERVER_INTERFACE_H
 
 #include <grpc/impl/codegen/grpc_types.h>
+#include <grpcpp/impl/codegen/byte_buffer.h>
 #include <grpcpp/impl/codegen/call_hook.h>
 #include <grpcpp/impl/codegen/completion_queue_tag.h>
 #include <grpcpp/impl/codegen/core_codegen_interface.h>
@@ -185,13 +186,13 @@ class ServerInterface : public internal::CallHook {
           notification_cq_(notification_cq),
           tag_(tag),
           request_(request) {
-      IssueRequest(registered_method, &payload_, notification_cq);
+      IssueRequest(registered_method, payload_.bbuf_ptr(), notification_cq);
     }
 
     bool FinalizeResult(void** tag, bool* status) override {
       if (*status) {
-        if (payload_ == nullptr ||
-            !SerializationTraits<Message>::Deserialize(payload_, request_)
+        if (!payload_.Valid() ||
+            !SerializationTraits<Message>::Deserialize(payload_.bbuf_ptr(), request_)
                  .ok()) {
           // If deserialization fails, we cancel the call and instantiate
           // a new instance of ourselves to request another call.  We then
@@ -219,7 +220,7 @@ class ServerInterface : public internal::CallHook {
     ServerCompletionQueue* const notification_cq_;
     void* const tag_;
     Message* const request_;
-    grpc_byte_buffer* payload_;
+    ByteBuffer payload_;
   };
 
   class GenericAsyncRequest : public BaseAsyncRequest {

--- a/include/grpcpp/impl/codegen/server_interface.h
+++ b/include/grpcpp/impl/codegen/server_interface.h
@@ -186,7 +186,7 @@ class ServerInterface : public internal::CallHook {
           notification_cq_(notification_cq),
           tag_(tag),
           request_(request) {
-      IssueRequest(registered_method, payload_.c_buffer_ptr(), notification_cq);
+      IssueRequest(registered_method, payload_.bbuf_ptr(), notification_cq);
     }
 
     ~PayloadAsyncRequest() {

